File: src/python/restx/components/BaseComponent.py
"""
RESTx: Sane, simple and effective data publishing and integration. 

Copyright (C) 2010   MuleSoft Inc.    http://www.mulesoft.com

This program is free software: you can redistribute it and/or modify 
it under the terms of the GNU General Public License as published by 
the Free Software Foundation, either version 3 of the License, or 
(at your option) any later version. 

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
GNU General Public License for more details. 

You should have received a copy of the GNU General Public License 
along with this program.  If not, see <http://www.gnu.org/licenses/>. 

"""

"""
Defines a base class for all components.

"""
# Python imports
import urllib, urllib2
from copy import deepcopy

#RESTx imports
import restx.settings as settings

from restx.core.parameter                  import *
from restx.storageabstraction.file_storage import FileStorage

from org.mulesoft.restx.util import Url

#
# Utility method.
#
def _change_params_to_plain_dict(param_dict):
    """
    Unwraps all the parameter definitions into a plain dictionary.
    
    Needed for browsing or accessing the meta info.
    
    @param param_dict: Dictionary of parameter definitions, with each
                       individual parameter represented by a ParameterDef object.
    @type param_dict:  dict or ParameterDef objects
    
    @return:           Plain dictionary.
    @rtype:            dict
    
    """
    d = dict()
    for name in param_dict.keys():
        d[name] = param_dict[name].as_dict()
    return d


class BaseComponent(object):
    """
    The base class for all code components.
    
    A component author only needs to override a few methods or data values
    to get started.
    
    Specifically:
    
        NAME
        PARAM_DEFINITION
        DESCRIPTION
        DOCUMENTATION
        SERVICES
        
    A component specifies 'sub-SERVICES'. To do so, implement any number of services
    methods. Then list those methods in the 'SERVICES' dictionary (use any name as
    key to a tuple containing the method itself as well as a doc string). Then your
    additional methods will be exposed as sub-services, which can be accessed
    like this: <resource_uri>/<sub_service_method>/....

    The name of the sub-service method is directly specified in the URI.
    
    """
    LANGUAGE         = "PYTHON"
    
    NAME             = ""
    """The name used to refer to this component. Also used to construct its URL. No spaces allowed."""
    PARAM_DEFINITION = dict()
    """The parameter definition in the form of a dictionary."""
    DESCRIPTION      = ""
    """A short, one line description."""
    DOCUMENTATION    = ""
    """Longer, man-page style documentation."""
    SERVICES         = None
    """A dictionary keying method name to docstring for exposed sub-service methods. May be left empty."""
    
    def __init__(self):
        self.__resource_name     = None
        self.__http_request      = None
        self.__base_capabilities = None
        
    def setBaseCapabilities(self, base_capabilities):
        self.__base_capabilities = base_capabilities

    def setResourceName(self, resource_name):
        self.__resource_name = resource_name
        
    def setRequest(self, request):
        self.__http_request = request
        
    def getRequestUri(self):
        return self.__http_request.getRequestURI()

    def getRequestHeaders(self):
        return self.__http_request.getRequestHeaders()
    
    def getMyResourceName(self):
        return self.__resource_name

    def getMyResourceUri(self):
        return "%s/%s" % (settings.PREFIX_RESOURCE, self.getMyResourceName())

    def getFileStorage(self, namespace=""):
        """
        Return a FileStorage object, which can be used to store data.

        Storage spaces for each resource are separated by resource name,
        this means that two resources cannot share their stored objects,
        even if they are of the same type.

        @param namespace:   A namespace that is used by this resource.
                            Per invocation a resource may chose to create
                            yet another resource namespace under (or within)
                            its inherent namespace.
        @type namespace:    string

        @return:            FileStorage object.

        """
        return self.__base_capabilities.getFileStorage(namespace)
    
    def httpSetCredentials(self, accountname, password):
        """
        The component author can set credentials for sites that require authentication.
        
        @param accountname:    Name of account
        @type accountname:     string
        
        @param password:       Password for account.
        @type password:        string
        
        """
        return self.__base_capabilities.httpSetCredentials(accountname, password)
    
    def httpGet(self, url, headers=None):
        """
        Accesses the specified URL.
        
        If credentials have been specified, they will be used in case
        of HTTP basic authentication.
        
        @param url:        The URL to be accessed.
        @type url:         string
        
        @param headers:    A dictionary of additional HTTP request headers.
        @type headers:     dict
        
        @return:           Status and data as tuple.
        @rtype:            tuple
        
        """
        res = self.__base_capabilities.httpGet(url, headers=headers)
        return (res.status, res.data)
    
    def httpPost(self, url, data, headers=None):
        """
        Send the specified data to the specified URL with the POST method.
        
        If credentials have been specified, they will be used in case
        of HTTP basic authentication.
        
        @param url:        The URL to be accessed.
        @type url:         string
        
        @param data:       The data to be sent to the URL.
        @type data:        string
        
        @param headers:    A dictionary of additional HTTP request headers.
        @type headers:     dict
        
        @return:           Status and data as tuple.
        @rtype:            tuple
        
        """
        res = self.__base_capabilities.httpPost(url, data, headers)
        return (res.status, res.data)

    def getMetaData(self):
        """
        Return meta data about this code.
        
        Contains name, description, documentation URI and parameter
        definitions as dictionary.
        
        @return: Meta info about the component.
        @rtype:  dict
        
        """
        d = dict(uri    = Url(self.getCodeUri()),
                 name   = self.getName(),
                 desc   = self.getDesc(),
                 doc    = Url(self.getCodeUri() + "/doc"),
                 params = _change_params_to_plain_dict(self.getParams()),
                 services = self._getServices()
                )
        #
        # There is also a set of resource meta parameters that always remain
        # the same. Just some of the defaults and descriptions may change
        # from component to component.
        #
        rp = dict(suggested_name = ParameterDef(PARAM_STRING,
                                                "Can be used to suggest the resource name to the server",
                                                required=True),
                  desc           = ParameterDef(PARAM_STRING,
                                                "Specifies a description for this new resource",
                                                required=False, default="A '%s' resource" % self.getName())
                 )
        d['resource_creation_params'] = _change_params_to_plain_dict(rp)
        return d
    
    def getName(self):
        """
        Return the name of the component.
        
        @return:  Name
        @rtype:   string
        
        """
        return self.NAME

    def getDesc(self):
        """
        Return the brief description string of the component.
        
        @return:  Description.
        @rtype:   string
        
        """
        return self.DESCRIPTION
    
    def getDocs(self):
        """
        Return the documentation text for this component.
        
        @return:  Documentation.
        @rtype:   string
        
        """
        return self.DOCUMENTATION

    def getCodeUri(self):
        """
        Return the URI for this component.
        
        @return: URI for the component.
        @rtype:  string
        
        """
        return settings.PREFIX_CODE + "/" + self.getName()
    
    def getParams(self):
        """
        Return the parameter definition for this component.
        
        @return: Parameter definition.
        @rtype:  dict (of ParameterDef objects)
        
        """
        return self.PARAM_DEFINITION

    
    #
    # -------------------------------------------------------------------------------------
    # Following are some methods that are used by the framework and that are not part
    # of the official component-API.
    # -------------------------------------------------------------------------------------
    #
    def _getServices(self, resource_base_uri=None):
        """
        Return a dictionary with the exposed services.
        
        Keyed by name, for each service we return the uri and short description
        and service-specific parameter definition.
        
        @param resource_base_uri:  The base URI for the resource in which the services are
                                   exposed. The base URI through which services are accessed
                                   is the one of the resource (not of this code). Therefore,
                                   if we want the URI of the services in the context of the
                                   resource then we can call this method here with the base
                                   URI of the resource passed in.
        @type resource_base_uri:   string
        
        @return:                   Dictionary with sub-service info.
        @rtype:                    dict
        
        """
        if resource_base_uri:
            base_uri = resource_base_uri
        else:
            # No base URI specified? Then we can get the service URIs relative to
            # the code's URI.
            base_uri = self.getCodeUri()
        if self.SERVICES:
            ret = dict()
            for name in self.SERVICES.keys():
                ret[name]  = deepcopy(self.SERVICES[name])  # That's a dictionary with params definitions and descs
                # Create proper dict representations of each parameter definition
                if 'params' in ret[name]:
                    for pname in ret[name]['params'].keys():
                        if type(ret[name]['params'][pname]) is ParameterDef:
                            # Need the type check since we may have constructed the
                            # representation from storage, rather than in memory.
                            # If it's from storage then we don't have ParameterDefs
                            # in this dictionary here, so we don't need to convert
                            # anything.
                            ret[name]['params'][pname] = ret[name]['params'][pname].as_dict()
                ret[name]['uri'] = Url(base_uri + "/" + name)
            return ret
            #return dict([(name, dict(uri=base_uri + "/" + name, desc=self.SERVICES[name])) for name in self.SERVICES.keys() ])
        else:
            return None

File: src/java/org/mulesoft/restx/component/BaseComponent.java
/*      
 *  RESTx: Sane, simple and effective data publishing and integration. 
 *  
 *  Copyright (C) 2010   MuleSoft Inc.    http://www.mulesoft.com 
 *  
 *  This program is free software: you can redistribute it and/or modify 
 *  it under the terms of the GNU General Public License as published by 
 *  the Free Software Foundation, either version 3 of the License, or 
 *  (at your option) any later version. 
 * 
 *  This program is distributed in the hope that it will be useful, 
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of 
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 *  GNU General Public License for more details. 
 * 
 *  You should have received a copy of the GNU General Public License 
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */ 


package org.mulesoft.restx.component;

import com.sun.net.httpserver.Headers;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.lang.annotation.Annotation;
import java.lang.reflect.*;
import java.math.BigDecimal;

import org.json.JSONException;

import org.mulesoft.restx.RestxHttpRequest;
import org.mulesoft.restx.ResourceAccessorInterface;
import org.mulesoft.restx.Settings;
import org.mulesoft.restx.component.api.*;
import org.mulesoft.restx.exception.RestxException;
import org.mulesoft.restx.parameter.*;
import org.mulesoft.restx.util.Url;
import org.mulesoft.restx.util.JsonProcessor;


public abstract class BaseComponent
{
    public final String                               LANGUAGE = "JAVA";
    public       ComponentDescriptor                  componentDescriptor = null;
    public       ResourceAccessorInterface            resourceAccessor;

    protected    HashMap<String, Object>              services;

    private      RestxHttpRequest                     httpRequest;
    private      String                               resourceName;
    private      BaseComponentCapabilities            baseCapabilities;
    
    private      boolean                              annotationsHaveBeenParsed = false;
    // We use the following to record the order of parameters as well as their
    // Java types for each service when we are parsing the annotations. Later,
    // this helps our service-method calling proxy to arrange the parameters in
    // the right order - since Java does not allow the **fkwargs notation of named
    // parameters - and also allows us to do the necessary type casting.
    private      HashMap<String, ArrayList<String>>   paramOrder;
    private      HashMap<String, ArrayList<Class<?>>> paramTypes;
    
    public BaseComponent()
    {
        this.resourceName     = null;
        this.httpRequest      = null;
        this.baseCapabilities = null;
    }
    
    public void setBaseCapabilities(BaseComponentCapabilities baseCapabilities)
    {
        this.baseCapabilities = baseCapabilities;
    }
    
    private ComponentDescriptor getComponentDescriptor() throws RestxException
    {
        if (componentDescriptor == null) {
            annotationParser();
        }
        return componentDescriptor;
    }
    
    public void setResourceName(String resourceName)
    {
        this.resourceName = resourceName;
    }
    
    public void setRequest(RestxHttpRequest request)
    {
        this.httpRequest = request;
    }
    
    public String getDocs()
    {
        return this.componentDescriptor.getDocs();
    }
    
    public String getRequestUri()
    {
        return httpRequest.getRequestURI();
    }
    
    public Headers getRequestHeaders()
    {
        return httpRequest.getRequestHeaders();
    }
    
    public HttpResult accessResource(String uri)
    {
        return accessResource(uri, null, null, HTTP.GET);
    }

    public HttpResult accessResource(String uri, String input)
    {
        return accessResource(uri, input, null, HTTP.GET);
    }

    public HttpResult accessResource(String uri, String input, Map<?,?> params)
    {
        return accessResource(uri, input, params, HTTP.GET);
    }
    
    public HttpResult accessResource(String uri, String input, Map<?,?> params, HttpMethod method)
    {
        return resourceAccessor.accessResourceProxy(uri, input, params, method);
    }
    
    public MakeResourceResult makeResource(String componentClassName, String suggestedResourceName,
                                           String resourceDescription, Map<?,?> resourceParameters) throws RestxException
    {
        return resourceAccessor.makeResourceProxy(componentClassName, suggestedResourceName, resourceDescription, resourceParameters);
    }
    
    private ParameterDef createParamDefType(Class<?> paramType, String desc,
                                            boolean required, String defaultVal)
    {
        ParameterDef pdef = null;
        if (paramType == String.class) {
            pdef = new ParameterDefString(desc, required, defaultVal);
        }
        else if (paramType == Integer.class ||  paramType == BigDecimal.class  ||
                 paramType == Double.class  ||  paramType == Float.class ||
                 paramType == int.class     ||  paramType == float.class ||
                 paramType == double.class)
        {
            pdef = new ParameterDefNumber(desc, required, new BigDecimal(defaultVal));
        }
        else if (paramType == Boolean.class) {
            pdef = new ParameterDefBoolean(desc, required, new Boolean(defaultVal));
        }
        return pdef;
    }
        
    public void annotationParser() throws RestxException
    {
        if (annotationsHaveBeenParsed  ||  componentDescriptor != null) {
            return;
        }
        Class<? extends BaseComponent> myclass = this.getClass();
        
        /*
         * Examine the class annotations to get information about the
         * component.
         */
        ComponentInfo ci = this.getClass().getAnnotation(ComponentInfo.class);
        if (ci == null) {
            throw new RestxException("Component does not have a ComponentInfo annotation");
        }
        componentDescriptor = new ComponentDescriptor(ci.name(), ci.description(), ci.doc());
        
        /*
         * Examine field annotations to identify resource creation
         * time parameters.
         */
        for (Field f: myclass.getFields()) {
            Parameter fa = f.getAnnotation(Parameter.class);
            if (fa != null) {
                String  paramName   = fa.name();
                String  paramDesc   = fa.desc();
                String  defaultVal  = null;
                boolean required    = true;
                
                // Check if we have a default value and set that one as well
                Default fad = f.getAnnotation(Default.class);
                if (fad != null) {
                    defaultVal = fad.value();
                    required   = false;
                }
                Class<?> ftype = f.getType();
                componentDescriptor.addParameter(paramName,
                                                 createParamDefType(ftype, paramDesc, required, defaultVal));
            }
        }
        
        /*
         * Examine the method annotations to identify service methods
         * and their parameters.
         */
        paramOrder = new HashMap<String, ArrayList<String>>();
        paramTypes = new HashMap<String, ArrayList<Class<?>>>();
        for (Method m: myclass.getMethods()) {
            if (m.isAnnotationPresent(Service.class)) {
                String            serviceName         = m.getName();
                Service           at                  = m.getAnnotation(Service.class);
                ServiceDescriptor sd                  = new ServiceDescriptor(at.description());
                Class<?>[]        types               = m.getParameterTypes();
                Annotation[][]    allParamAnnotations = m.getParameterAnnotations();
                int i = 0;
                ArrayList<String> positionalParams = new ArrayList<String>();
                for (Annotation[] pa : allParamAnnotations) {
                    String  name       = null;
                    String  desc       = null;
                    boolean required   = true;
                    String  defaultVal = null;
                    for (Annotation a: pa) {
                        if (a.annotationType() == Parameter.class) {
                            name = ((Parameter)a).name();
                            desc = ((Parameter)a).desc();
                            if (!paramOrder.containsKey(serviceName)) {
                                paramOrder.put(serviceName, new ArrayList<String>());
                                paramTypes.put(serviceName, new ArrayList<Class<?>>());
                            }
                            if (((Parameter)a).positional()) {
                                positionalParams.add(name);
                            }
                            paramOrder.get(serviceName).add(name);
                            paramTypes.get(serviceName).add(types[i]);
                        }
                        else if (a.annotationType() == Default.class) {
                            defaultVal = ((Default)a).value();
                            required = false;                            
                        }
                    }
                    if (name != null) {                     
                        sd.addParameter(name, createParamDefType(types[i], desc, required, defaultVal));
                    }
                    ++i;
                }
                if (!positionalParams.isEmpty()) {
                    sd.setPositionalParameters(positionalParams);
                }
                componentDescriptor.addService(m.getName(), sd);
            }
        }
        annotationsHaveBeenParsed = true;
    }
    
    public HashMap<String, ArrayList<String>> getParameterOrder() throws RestxException
    {
        annotationParser();

        // Returns the order in which parameters were defined
        return paramOrder;
    }
    
    public HashMap<String, ArrayList<Class<?>>> getParameterTypes() throws RestxException
    {
        annotationParser();

        // Returns the order in which parameters were defined
        return paramTypes;
    }
    
    public String getMyResourceName()
    {
        return resourceName;
    }
    
    public String getMyResourceUri()
    {
        return Settings.PREFIX_RESOURCE + "/" + getMyResourceName();
    }
    
    public FileStore getFileStorage()
    {
        return baseCapabilities.getFileStorage();
    }

    public FileStore getFileStorage(String namespace)
    {
        return baseCapabilities.getFileStorage(namespace);
    }
    
    public void httpSetCredentials(String accountName, String password)
    {
        baseCapabilities.httpSetCredentials(accountName, password);
    }
    
    public HttpResult httpGet(String url)
    {
        return baseCapabilities.httpGet(url);
    }
    
    public HttpResult httpGet(String url, HashMap<String, String> headers)
    {
        return baseCapabilities.httpGet(url, headers);
    }
    
    public HttpResult httpPost(String url, String data)
    {
        return baseCapabilities.httpPost(url, data);
    }
    
    public HttpResult httpPost(String url, String data, Map<String, String> headers)
    {
        return baseCapabilities.httpPost(url, data, headers);
    }

    // JSON processing
    public Object fromJsonStr(String str) throws RestxException
    {
        try {
            return JsonProcessor.loads(str);
        }
        catch (JSONException e) {
            throw new RestxException("Could not de-serialize data: " + e.getMessage());
        }
    }

    public String toJsonStr(Object obj) throws RestxException
    {
        try {
            return JsonProcessor.dumps(obj);
        }
        catch (JSONException e) {
            throw new RestxException("Could not serialize data: " + e.getMessage());
        }
    }
    
    private HashMap<String, Object> changeParamsToPlainDict(Map<String, ParameterDef> paramDict)
    {
        HashMap<String, Object> d = new HashMap<String, Object>();
        for (String name: paramDict.keySet()) {
            d.put(name, paramDict.get(name).asDict());
        }
        return d;
    }
    
    public HashMap<String, Object> getMetaData() throws RestxException
    {
        annotationParser();

        HashMap<String, Object> d = new HashMap<String, Object>();
        
        d.put("uri",      new Url(getCodeUri()));
        d.put("name",     getName());
        d.put("desc",     getDesc());
        d.put("doc",      getCodeUri() + "/doc");
        d.put("params",   changeParamsToPlainDict(componentDescriptor.getParamMap()));
        d.put("services", _getServices(null));
        
        HashMap<String, ParameterDef> rp = new HashMap<String, ParameterDef>();
        rp.put("suggested_name", new ParameterDefString("Can be used to suggest the resource name to the server",
                                                        true, ""));
        rp.put("desc",           new ParameterDefString("Specifies a description for this new resource",
                                                        false, "A '" + getName() + "' resource")); 

        d.put("resource_creation_params", changeParamsToPlainDict(rp));
        
        return d;
    }

    public HashMap<String, ParameterDef> getParams() throws RestxException
    {
        if (componentDescriptor == null) {
            annotationParser();
        }
        return componentDescriptor.getParamMap();
    }
    
    public String getName() throws RestxException
    {
        return getComponentDescriptor().getName();
    }
    
    public String getDesc() throws RestxException
    {
        return getComponentDescriptor().getDesc();
    }
    
    public String getDoc() throws RestxException
    {
        return getComponentDescriptor().getDocs();
    }
    
    public String getCodeUri()
    {
        String name;
        try {
            name = getName();
        }
        catch (Exception e) {
            name = "";
        }
        return Settings.PREFIX_CODE + "/" + name;
    }
    
    /*
     * Following are some methods that are used by the framework and that are not part
     * of the official component-API.
     */
    
    /*
     * Return a dictionary of all defined services. resourceBaseUri may be set to null,
     * in which case all service URLs are relative to the code URL of the component.
     */
    public HashMap<String, Object> _getServices(String resourceBaseUri) throws RestxException
    {
        annotationParser();

        // Get the base URI for all services. If no resource base URI
        // was defined (can happen when we just look at code meta data)
        // then we use the code base URI instead.
        String baseUri;        
        if (resourceBaseUri == null) {
            baseUri = getCodeUri();
        }
        else {
            baseUri = resourceBaseUri;
        }
        
        // Create a map of service descriptions.
        if (componentDescriptor.getServicMap() != null  &&  !componentDescriptor.getServicMap().isEmpty()) {            
            services = componentDescriptor.getServicesAsPlainDict();
            HashMap<String, Object> ret = new HashMap<String, Object>();
            for (String name: services.keySet()) {
                HashMap<String, Object> thisService = (HashMap<String, Object>)services.get(name);
                thisService.put("uri", new Url(baseUri + "/" + name));
                ret.put(name, thisService);
                HashMap<String, Object> params = (HashMap<String, Object>)thisService.get("params");
                if (params != null) {
                    for (String pname: params.keySet()) {
                        Object param = params.get(pname);
                        if (param instanceof ParameterDef) {
                            // Need the type check since we may have constructed the
                            // representation from storage, rather than in memory.
                            // If it's from storage then we don't have ParameterDefs
                            // in this dictionary here, so we don't need to convert
                            // anything.
                            param = ((ParameterDef)param).asDict();
                            params.put(pname, param);
                        }
                    }
                }
            }
            
            return ret;
        }
        else {
            // No services defined? Nothing to return...
            return null;
        }
    }
 }


